storage/index.go

```diff
@@ -207,156 +205,144 @@ func newNamespaceIndexWithOptions(
 	// by primitives internal to it.
 	i.state.RLock()
 	defer i.state.RUnlock()
-
 	if !i.isOpenWithRLock() {
 		// NB(prateek): deliberately skip calling any of the `OnIndexFinalize` methods
 		// on the provided inserts to terminate quicker during shutdown.
-		i.metrics.InsertAfterClose.Inc(int64(len(inserts)))
+		return
 	}
 
-	// we sort the inserts by which block they're applicable for, and do the inserts
-	// for each block.
-	writesByBlockStart := index.WriteBatchEntryByBlockStart(inserts)
-	sort.Sort(writesByBlockStart)
-	writesByBlockStart.ForEachBlockStart(i.writeBatchForBlockStartWithRLock)
+	now := i.nowFn()
+	futureLimit := now.Add(1 * i.bufferFuture)
+	pastLimit := now.Add(-1 * i.bufferPast)
+	writeBatchFn := i.writeBatchForBlockStartWithRLock
+	for _, batch := range batches {
+		// Ensure timestamp is not too old/new based on retention policies and that
+		// doc is valid.
+		batch.ForEach(func(idx int, entry index.WriteBatchEntry,
+			d doc.Document, _ index.WriteBatchEntryResult) {
+			if !futureLimit.After(entry.Timestamp) {
+				batch.MarkUnmarkedEntryError(m3dberrors.ErrTooFuture, idx)
+				return
+			}
+
+			if !entry.Timestamp.After(pastLimit) {
+				batch.MarkUnmarkedEntryError(m3dberrors.ErrTooPast, idx)
+				return
+			}
+		})
+
+		// we sort the inserts by which block they're applicable for, and do the inserts
+		// for each block, making sure to not try to insert any entries already marked
+		// with a result.
+		batch.ForEachUnmarkedBatchByBlockStart(writeBatchFn)
+	}
 }
 
 func (i *nsIndex) writeBatchForBlockStartWithRLock(
-	blockStart xtime.UnixNano, inserts []index.WriteBatchEntry,
+	blockStart time.Time, batch *index.WriteBatch,
 ) {
 	// ensure we have an index block for the specified blockStart.
-	block, err := i.ensureBlockPresentWithRLock(blockStart.ToTime())
+	block, err := i.ensureBlockPresentWithRLock(blockStart)
 	if err != nil {
-		for _, insert := range inserts {
-			insert.OnIndexSeries.OnIndexFinalize()
-		}
+		batch.MarkUnmarkedEntriesError(err)
 		i.logger.WithFields(
 			xlog.NewField("blockStart", blockStart),
-			xlog.NewField("numWrites", len(inserts)),
+			xlog.NewField("numWrites", batch.Len()),
 			xlog.NewField("err", err.Error()),
 		).Error("unable to write to index, dropping inserts.")
-		i.metrics.AsyncInsertErrors.Inc(int64(len(inserts)))
+		i.metrics.AsyncInsertErrors.Inc(int64(batch.Len()))
 		return
 	}
 
 	// i.e. we have the block and the inserts, perform the writes.
-	result, err := block.WriteBatch(inserts)
+	result, err := block.WriteBatch(batch)
+
 	// NB: we don't need to do anything to the OnIndexSeries refs in `inserts` at this point,
 	// the index.Block WriteBatch assumes responsibility for calling the appropriate methods.
 	if numErr := result.NumError; numErr != 0 {
 		i.metrics.AsyncInsertErrors.Inc(numErr)
 	}
 	if err != nil {
-		i.logger.Errorf("unable to write to index, dropping inserts. [%v]", err)
+		i.logger.Errorf("error writing to index block: %v", err)
```

Comment:
super nit: could you filter out any `ErrDuplicateID` from here?

Reply:
I see you are doing this =]